File: web-app/src/types/schedule.ts
/** Mirrors `agentic_schedules` (backend `schedule::Model`, serde snake_case). */
export type ScheduleTargetKind = "workflow" | "airway" | "agent" | "monitor_scan";

export interface Schedule {
  id: string;
  project_id: string | null;
  branch_id: string | null;
  name: string;
  target_kind: ScheduleTargetKind;
  /** workflow_ref / pipeline_ref / agent_id, workspace-relative. */
  target_ref: string;
  /** Required when `target_kind === "agent"`, null otherwise. */
  question: string | null;
  variables: Record<string, unknown> | null;
  cron_expr: string;
  timezone: string;
  enabled: boolean;
  /** ISO 8601. */
  next_run_at: string;
  last_fired_at: string | null;
  last_run_id: string | null;
  /** Most recent fire/seed/cron failure; null once a fire succeeds. */
  last_error: string | null;
  /**
   * Cumulative count of cron occurrences that were silently skipped
   * because the scheduler tick didn't run in time (server downtime,
   * deploy gap). Policy is "run-once-then-resume" — only the first
   * missed slot fires; everything else past it is counted here.
   * Never decremented; purely audit.
   */
  missed_runs: number;
  /**
   * Timestamp (ISO 8601) of the most recent tick that detected a
   * catch-up. `null` when `missed_runs == 0`.
   */
  last_missed_at: string | null;
  created_at: string;
  updated_at: string;
}

/** Create/update payload (backend `ScheduleInput`). */
export interface ScheduleInput {
  name: string;
  target_kind: ScheduleTargetKind;
  target_ref: string;
  /** Required when `target_kind === "agent"`. Backend rejects an empty
   *  question for agent schedules; ignored for workflow / airway. */
  question?: string | null;
  variables?: Record<string, unknown> | null;
  cron_expr: string;
  /** Defaults to "UTC" server-side. */
  timezone?: string;
  /** Defaults to true server-side. */
  enabled?: boolean;
}

export interface RunNowResponse {
  run_id: string;
}

/** Body of `POST /agentic-schedules/:id/backfill`. */
export interface BackfillInput {
  /** Inclusive lower bound (ISO 8601 / RFC 3339). */
  from: string;
  /** Inclusive upper bound (ISO 8601 / RFC 3339); must be > `from`. */
  to: string;
  /** Throttle hint: `"sequential"` | `"<N>"` | `"all"`. Advisory in v1. */
  concurrency?: string;
}

export interface BackfillResponse {
  run_ids: string[];
  /** Cron occurrences enumerated; may exceed `run_ids.length` on partial seed. */
  planned: number;
}

File: crates/agentic/runtime/src/lifecycle/entity/schedule.rs
//! Cron schedule row (Phase 2). A user-defined recurring trigger that the
//! scheduler tick fires by seeding a `TaskScope::Global` run.

use sea_orm::entity::prelude::*;
use serde::Serialize;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize)]
#[sea_orm(table_name = "agentic_schedules")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: String,
    /// Owning workspace (multi-tenant scope). Plain Uuid, no FK per the
    /// agentic cross-domain loose-references rule; app-level cleanup on
    /// workspace delete. New rows always set this via the CRUD handlers;
    /// any pre-existing rows backfill to the nil UUID and are inert.
    pub workspace_id: Uuid,
    /// Legacy/optional project + branch identifiers (unused; superseded
    /// by `workspace_id` for scoping). Kept for back-compat / future
    /// per-branch scoping work.
    pub project_id: Option<String>,
    pub branch_id: Option<String>,
    /// User-facing label.
    pub name: String,
    /// `"workflow"` | `"airway"` | `"agent"`.
    pub target_kind: String,
    /// `workflow_ref` / `pipeline_ref` / `agent_id`, workspace-relative.
    pub target_ref: String,
    /// Free-text question for agent schedules. Required when
    /// `target_kind = "agent"`; ignored for workflow / airway. Stored as
    /// `Option<String>` so existing rows (and the two other target kinds)
    /// leave it NULL.
    pub question: Option<String>,
    /// Variables passed to the seed fn (JSON object).
    pub variables: Option<serde_json::Value>,
    /// Standard cron expression.
    pub cron_expr: String,
    /// IANA timezone the cron expression is evaluated in.
    pub timezone: String,
    pub enabled: bool,
    /// The CAS coordination point: the next due time. Advanced atomically
    /// on fire so exactly one replica's tick wins.
    pub next_run_at: DateTimeWithTimeZone,
    /// Observability: when the schedule last fired and the run it seeded.
    pub last_fired_at: Option<DateTimeWithTimeZone>,
    pub last_run_id: Option<String>,
    /// Most recent fire/seed failure (bad cron, missing target, seed
    /// error). `NULL` once a fire succeeds.
    pub last_error: Option<String>,
    /// Cumulative count of cron occurrences that fell between
    /// successive `next_run_at` advances — i.e., the number of slots
    /// silently skipped while the server was down. Incremented by the
    /// scheduler tick at catch-up time; never decremented (purely
    /// audit). 0 in steady state when every tick interval runs.
    pub missed_runs: i32,
    /// Timestamp of the most recent tick that detected a catch-up.
    /// `NULL` at row creation; stamped only when [`missed_runs`] increments.
    pub last_missed_at: Option<DateTimeWithTimeZone>,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
